storage/src/main/java/tech/pegasys/artemis/storage/server/rocksdb/core/RocksDbInstance.java

```diff
@@ -0,0 +1,250 @@
+
+    private Transaction(
+        final TransactionDB db,
+        final ColumnFamilyHandle defaultHandle,
+        final ImmutableMap<RocksDbColumn<?, ?>, ColumnFamilyHandle> columnHandles) {
+      this.defaultHandle = defaultHandle;
+      this.columnHandles = columnHandles;
+      this.writeOptions = new WriteOptions();
+      this.rocksDbTx = db.beginTransaction(writeOptions);
+    }
+
+    public <T> void put(RocksDbVariable<T> variableType, T value) {
+      assertOpen();
+      final byte[] serialized = variableType.getSerializer().serialize(value);
+      try {
+        rocksDbTx.put(defaultHandle, variableType.getId().toArrayUnsafe(), serialized);
+      } catch (RocksDBException e) {
+        throw new DatabaseStorageException("Failed to put variable", e);
+      }
+    }
+
+    public <K, V> void put(RocksDbColumn<K, V> column, K key, V value) {
+      assertOpen();
+      final byte[] keyBytes = column.getKeySerializer().serialize(key);
+      final byte[] valueBytes = column.getValueSerializer().serialize(value);
+      final ColumnFamilyHandle handle = columnHandles.get(column);
+      try {
+        rocksDbTx.put(handle, keyBytes, valueBytes);
+      } catch (RocksDBException e) {
+        throw new DatabaseStorageException("Failed to put column data", e);
+      }
+    }
+
+    public <K, V> void put(RocksDbColumn<K, V> column, Map<K, V> data) {
+      assertOpen();
+      final ColumnFamilyHandle handle = columnHandles.get(column);
+      for (Entry<K, V> kvEntry : data.entrySet()) {
+        final byte[] key = column.getKeySerializer().serialize(kvEntry.getKey());
+        final byte[] value = column.getValueSerializer().serialize(kvEntry.getValue());
+        try {
+          rocksDbTx.put(handle, key, value);
+        } catch (RocksDBException e) {
+          throw new DatabaseStorageException("Failed to put column data", e);
+        }
+      }
+    }
+
+    public <K, V> void delete(RocksDbColumn<K, V> column, K key) {
+      assertOpen();
+      final ColumnFamilyHandle handle = columnHandles.get(column);
+      try {
+        rocksDbTx.delete(handle, column.getKeySerializer().serialize(key));
+      } catch (RocksDBException e) {
+        throw new DatabaseStorageException("Failed to delete key", e);
+      }
+    }
+
+    public void commit() {
+      assertOpen();
+      try {
+        this.rocksDbTx.commit();
+        close();
+      } catch (RocksDBException e) {
+        rollback();
+        throw new DatabaseStorageException("Failed to commit transaction", e);
+      }
+    }
+
+    public void rollback() {
+      assertOpen();
+      try {
+        this.rocksDbTx.rollback();
+        close();
+      } catch (RocksDBException e) {
+        throw new DatabaseStorageException("Failed to rollback transaction", e);
+      }
+    }
+
+    private void assertOpen() {
+      if (closed.get()) {
+        throw new IllegalStateException("Attempt to update a closed transaction");
+      }
```

Comment:
nit: could just use `Preconditions.checkState(closed.get(), "Attempt to update a closed transaction")`